daemon/graphdriver/devmapper/deviceset.go

```diff
@@ -841,11 +841,69 @@ func (devices *DeviceSet) createRegisterDevice(hash string) (*devInfo, error) {
 	return info, nil
 }
 
-func (devices *DeviceSet) createRegisterSnapDevice(hash string, baseInfo *devInfo, size uint64) error {
-	if err := devices.poolHasFreeSpace(); err != nil {
+func (devices *DeviceSet) takeSnapshot(hash string, baseInfo *devInfo, size uint64) error {
+	var (
+		doSuspend      bool
+		devinfo        *devicemapper.Info
+		deferredRemove = true
+		err            error
+	)
+
+	if err = devices.poolHasFreeSpace(); err != nil {
+		return err
+	}
+
+	if devices.deferredRemove {
+		devinfo, err = devicemapper.GetInfoWithDeferred(baseInfo.Name())
+		if err != nil {
+			return err
+		}
+	} else {
+		deferredRemove = false
+		devinfo, err = devicemapper.GetInfo(baseInfo.Name())
+		if err != nil {
+			return err
+		}
+	}
+
+	if devinfo != nil && devinfo.DeferredRemove == 0 {
+		deferredRemove = false
+	}
+
```

Comment:
I think this check can go inside first block if if condition. Right after GetInfoWithDeferred().